### PR TITLE
Hyperparameter tuning

### PR DESCRIPTION
**Pull request by Tole-k**

### CODE REVIEW

**Review by Dawid64 — COMMENTED** [review]
So, as far as I see we will need to do much more work with devices, for single/multi cpu/gpu (maybe even qpu (just kidding)), but for now I would stay with CPU, and try not to much with other architectures (just for first iteration, we can add support to different devices in 0.2 update, as it's important, but not obligatory for now.
The second thing are hyper parameters, code is very messy and much harder to code around if most of it is just one list, I propose or make it smaller by using function(arg1, arg2, arg3), to make size of this list smaller, maybe consider putting it outside of the class and reference it through variable (it would be the best to put it at the end of code then), or even better, just put it in json file with "XGBoost: [...]", one big json file should be enough, may even speed the code by a bit (as python interpreter wouldn't to have large code files), but it's mostly cosmetic change.

File: quick_ai/algorithms/neural_network.py [hunk 1]
```diff
@@ -190,32 +252,34 @@ def create_neural_network(self, input_size: int, output_size: int, hidden_sizes:
...
+        return self.model.predict(guess)
+
+
+if __name__ == '__main__':
+    params = NNFactory.optimize_hyperparameters(
+        get_iris(), 'classification', 'cuda')
+    print(params)
+    model = NNFactory().create_neural_network(**params)
+    model.train(*get_iris())
+    print(model.predict(get_iris()[0]))
```
[the remark this answers is not in the document]
Comment (Tole-k) [reply quick_ai/algorithms/neural_network.py:285]: Yep, that's only to make sure everything works as expected

**Comment by Tole-k** [comment]
> So, as far as I see we will need to do much more work with devices, for single/multi cpu/gpu (maybe even qpu (just kidding)), but for now I would stay with CPU, and try not to much with other architectures (just for first iteration, we can add support to different devices in 0.2 update, as it's important, but not obligatory for now. The second thing are hyper parameters, code is very messy and much harder to code around if most of it is just one list, I propose or make it smaller by using function(arg1, arg2, arg3), to make size of this list smaller, maybe consider putting it outside of the class and reference it through variable (it would be the best to put it at the end of code then), or even better, just put it in json file with "XGBoost: [...]", one big json file should be enough, may even speed the code by a bit (as python interpreter wouldn't to have large code files), but it's mostly cosmetic change.

Regarding the first point, I wanted to design it the way so that you can enable cuda computations in algorithms supporting it with a simple change of 1 parameter, but never to force it or make mandatory, leaving cpu as the default everywhere, I would argue that it is a Valuable  addition even so early into the development, at least to make testing pieces of code less time consuming

**Comment by Dawid64** [comment]
> > So, as far as I see we will need to do much more work with devices, for single/multi cpu/gpu (maybe even qpu (just kidding)), but for now I would stay with CPU, and try not to much with other architectures (just for first iteration, we can add support to different devices in 0.2 update, as it's important, but not obligatory for now. The second thing are hyper parameters, code is very messy and much harder to code around if most of it is just one list, I propose or make it smaller by using function(arg1, arg2, arg3), to make size of this list smaller, maybe consider putting it outside of the class and reference it through variable (it would be the best to put it at the end of code then), or even better, just put it in json file with "XGBoost: [...]", one big json file should be enough, may even speed the code by a bit (as python interpreter wouldn't to have large code files), but it's mostly cosmetic change.
> 
> Regarding the first point, I wanted to design it the way so that you can enable cuda computations in algorithms supporting it with a simple change of 1 parameter, but never to force it or make mandatory, leaving cpu as the default everywhere, I would argue that it is a valuable addition even so early into the development, at least to make testing pieces of code less time-consuming

The problem is that by not-supported I mean not handled by pip at all, there is not build version for macOS, so macOS user's will not be able to use library, we would need to make import conditional and not mandatory, but I'm not sure if this would be optimal solution. I think that for now it would be the best to make it hard-coded/exclude it, as to make the whole interface work properly it would take too a lot of time

**Comment by Tole-k** [comment]
> > > So, as far as I see we will need to do much more work with devices, for single/multi cpu/gpu (maybe even qpu (just kidding)), but for now I would stay with CPU, and try not to much with other architectures (just for first iteration, we can add support to different devices in 0.2 update, as it's important, but not obligatory for now. The second thing are hyper parameters, code is very messy and much harder to code around if most of it is just one list, I propose or make it smaller by using function(arg1, arg2, arg3), to make size of this list smaller, maybe consider putting it outside of the class and reference it through variable (it would be the best to put it at the end of code then), or even better, just put it in json file with "XGBoost: [...]", one big json file should be enough, may even speed the code by a bit (as python interpreter wouldn't to have large code files), but it's mostly cosmetic change.
> > 
> > Regarding the first point, I wanted to design it the way so that you can enable cuda computations in algorithms supporting it with a simple change of 1 parameter, but never to force it or make mandatory, leaving cpu as the default everywhere, I would argue that it is a valuable addition even so early into the development, at least to make testing pieces of code less time-consuming
> 
> The problem is that by not-supported I mean not handled by pip at all, there is not build version for macOS, so macOS user's will not be able to use library, we would need to make import conditional and not mandatory, but I'm not sure if this would be optimal solution. I think that for now it would be the best to make it hard-coded/exclude it, as to make the whole interface work properly it would take too a lot of time

Is this about cupy package which fails to import? If so then I believe it is used only in one algorithm so we could import it only if there was device='cuda' parameter passed and resolve this issue that way, or is there something else which doesn't work on macOS ? I don't have access so I have no way of checking

**Comment by Dawid64** [comment]
> Is this about cupy package which fails to import? If so then I believe it is used only in one algorithm so we could import it only if there was device='cuda' parameter passed and resolve this issue that way, or is there something else which doesn't work on macOS ? I don't have access so I have no way of checking

It would be the minimum, Many libraries have special packets for gpu's and so on, cuda also should not work, but for now it's only cupy, so making import dependent should be enough

**Pull request merged**